content/binary-api.md

```diff
@@ -0,0 +1,227 @@
+  protected def d: Int = ...
+  inline def foo: Int = a + b + c + d
+}
+~~~
+before inlining the compiler will generate the accessors for inlined definitions
+~~~ scala
+class C {
+  @binaryAPIAccessor private def a: Int = ...
+  private def b: Int = ...
+  @binaryAPI protected def c: Int = ...
+  protected def d: Int = ...
+  final def C$inline$a: Int = ... // generated by `@binaryAPIAccessor`
+  final def C$inline$b: Int = ... // warn: `b` should be annotated with `@binaryAPIAccessor` + migration code
+  final def C$inline$d: Int = ... // warn: `d` should be annotated with `@binaryAPI` + migration code
+  inline def foo: Int = C$inline$a + C$inline$b + c + C$inline$d
+}
+~~~
+
+### Specification
+
+We must add `binaryAPI` and `binaryAPIAccessor` to the standard library.
+
+```scala
+package scala.annotation
+
+final class binaryAPI extends scala.annotation.StaticAnnotation
+final class binaryAPIAccessor extends scala.annotation.StaticAnnotation
+```
+
+#### `@binaryAPI` annotation
+
+* Only valid on `def`, `val`, `lazy val`, `var`, `object`, and `given`.
+* TASTy will contain references to non-public definitions that are out of scope but `@binaryAPI`. TASTy already allows those references.
+* Annotated definition will be public in the generated bytecode. Definitions should be made public as early as possible in the compiler phases, as this can remove the need to create other accessors. It should be done after we check the accessibility of references.
+
+
+#### `@binaryAPIAccessor` annotation
+
+* Only valid on `def`, `val`, `lazy val`, `var`, `object`, and `given`.
+* An public accessor will be generated for the annotated definition. This accessor will be named `<fullClassName>$$inline$<definitionName>`.
+
+#### Inline
+
+* Inlining will not require the generation of an inline accessor for binary APIs.
+* Inlining will not require the generation of a new inline accessor, it will use the binary API accessors.
+* The user will be warned if a new inline accessor is automatically generated.
+  The message will suggest `@binaryAPI` or `@binaryAPIAccessor` and how to fix potential incompatibilities.
+  In a future version, these will become an error.
+
+### Compatibility
+
+The introduction of the `@binaryAPI` and `@binaryAPIAccessor` do not introduce any binary incompatibility.
+
+Using references to `@binaryAPI` and `@binaryAPIAccessor` in inline code can cause binary incompatibilities. These incompatibilities are equivalent to the ones that can occur due to the unsoundness we want to fix. When migrating to binary APIs, the compiler will show the implementation of accessors that the users need to add to keep binary compatibility with pre-binaryAPI code.
+
+A definition can be both `@binaryAPI` and `@binaryAPIAccessor`. This would be used to indicate that the definition used to be private, but now we want to publish it as public. The definition would become public, and the accessor would be generated for binary compatibility.
```

Comment:
> the compiler will show the implementation of accessors that the users need to add to keep binary compatibility

...

> the accessor would be generated for binary compatibility


IIUC, currently accessors are generated next to the inline method, whereas `@binaryAPIAccessor` wil generate an accessor next to the annotated private method, so that migration cannot be made in a binary compatbile manner - rigtht?

Reply:
> currently accessors are generated next to the inline method

This is the case if the inline method is in a different class from the accessed member.

On the other hand, if the inline method is in the class the accessor will be placed in the class that defines the accessed members, similar to `@binaryAPIAccessor`. In some cases, these accessors are the same and we avoid binary incompatibilities in the migration from auto-generated to `@binaryAPIAccessor` (example 1).

> @binaryAPIAccessor wil generate an accessor next to the annotated private method, so that migration cannot be made in a binary compatbile manner?

In general, the migration can't be made binary-compatible. We chose to reuse the accessor name `<fullClassName>$$inline$<definitionName>` to be binary compatible with some auto-generated accessors. The accessors that are binary compatible are the ones in a non-final class that are generated by an inline method in that class (example 1).

#### Example 1 (compatible):
```scala
class A(x: Int):
  inline def y = x + 1 // inline def y = A$$inline$x + 1
  // def A$$inline$x = x // auto-generated
```
migrates to
```scala
class A(@binaryAPIAccessor x: Int):
  inline def y = x + 1 // inline def y = A$$inline$x + 1
  // def A$$inline$x = x // generated by @binaryAPIAccessor
```

#### Example 2 (incompatible):
```scala
final class A(x: Int):
  inline def y = x + 1 // inline def y = inline$x + 1
  // def inline$x = x // auto-generated
```
migrates to
```scala
class A(@binaryAPIAccessor x: Int):
  inline def y = x + 1 // inline def y = A$$inline$x + 1
  def inline$x = x // user needs to add this. Migration warning shows tells the users shows this code
  // def A$$inline$x = x // generated by @binaryAPIAccessor
```

#### Example 3 (incompatible):
```scala
package p
final class A(private[p] x: Int)
```
```scala
package p
inline def f(a: A): Int = a.x // inline def f(a: A): Int = A$$inline$x(a) 
// def A$$inline$x(a: A) = a.x // auto-generated
```
migrates to
```scala
package p
final class A(@binaryAPI private[p] x: Int)
```
```scala
package p
inline def f(a: A): Int = a.x // inline def f(a: A): Int = a.x
def A$$inline$x(a: A) = a.x // user needs to add this. Migration warning shows tells the users shows this code
```

Reply:
> In general, the migration can't be made binary-compatible

In that case, it's maybe not worth introducing both `@binaryAPI` and `@binaryAPIAccessor`? Is it always possible to start using `@binaryAPI` and maintain binary compatibility by manually adding the accessors that were previously generated by the compiler?

Reply:
The issue with that approach is that users would need to change all their `private/private[this]` members that are accessed from inline methods `protected/private[Cls]/pribate[pack]`. This can cause source incompatibilities.

Reply:
I don't understand, I'm sorry. Where am I wrong?
  - Currently, if a `private` method is accessed in an `inline def`, the compiler generates an accessor to the private method
  - The user can annotate the method `@binaryAPI` (but keep it private in source), which makes it public in bytecode, and the accessor is no longer generated
  - To maintain binary compatibility, the user can write the method that was previously generated into the source code

Reply:
> The user can annotate the method @binaryAPI (but keep it private in source), which makes it public in bytecode, and the accessor is no longer generated

[`@binaryAPI` cannot be placed on private/private[this] APIs](https://github.com/scala/improvement-proposals/pull/58/files#diff-fe3d0b5bf9913b4901dab94b4dc8046f1abb7a37eb3884c783125bda9f3f868dR65). This is because that change could be a source or binary incompatible change. For example:
```scala
class A:
  @binaryAPI private val x: Int = 1 // should not be allowed
  def f: Int = x
class B extends A:
  val x: Int = 2
```
If we make `x` public in the bytecode, we would accidentally make `B.x` override `A.x` which would change the semantics of a private definition. Or we would have to flag them as overrides of private members, which implies that we would have to rename anything that would clash with a `@binaryAPI private`. In this case we would not be able to use the name `x` in `B` anymore.

Reply:
Ah, right, thanks. Was name-mangling considered?

"Only add `@binaryAPIAccessor`" could also be an option. The concern

> The drawback is that we would add code size and runtime overhead to all uses of this feature

is maybe minor? Accessors are very common in Scala, why would these ones be problematic?

Reply:
> "Only add @binaryAPIAccessor" could also be an option. 

If we do that we would not cover 2 of the motivations: _Removing deprecated APIs_ and _No way to inline reference to private constructors_. Those use cases could be supported later with the addition of `@binaryAPI`.

If we add `@binaryAPI` later there is also a migration concern. In a first step users would add `@binaryAPIAccessor` to all the definitions, but then when `@binaryAPI` they might also want to add that one to access the member directly. This would force them to have both `@binaryAPIAccessor` and `@binaryAPI`. If we add both at the same time, then they would be able to only use `@binaryAPI`.

> Ah, right, thanks. Was name-mangling considered?

I assume with `@binaryAPIAccessor` of private members. The issue is that we would need to mangle the names of private members but not public/protected members. This implies that if the user changes a private member into a public one the accessor would not be generated anymore. This is the exact same issue we have right now but triggered in a different way.

Reply:
Thanks for the details. I don't want to be annoying and keep discussing details here, the discussion should move elsewhere?

I think introducing two annotations is bad. Users will have to pick one or the other, they have to know how they both work and how using them affects binary compatibility.

Reply:
I do not see a scenario where a user would need to think about which annotation to choose. In the scenario of _Removing deprecated APIs_ they can only use `@binaryAPI`. In all other scenarios with inlining, the users would be told which annotation they need to use.